protocol-select/README.md

```diff
@@ -0,0 +1,541 @@
+    - [FAQ](#faq)
+
+## Introduction
+
+_Protocol Select_ is a protocol negotiation protocol. It is aimed at negotiating
+libp2p protocols on connections and streams. It replaces the _[Multistream
+Select]_ protocol.
+
+### Improvements over _[Multistream Select]_
+
+- **Downgrade attacks** and **censorship resistance**
+
+  Given that **[Multistream Select]** negotiates a connection's security
+  protocol unencrypted and unauthenticated it is prone to [downgrade attack]s.
+  In addition, a man-in-the-middle can detect that a given connection is used
+  to carry libp2p traffic, allowing attackers to censor such connections.
+
+  **Protocol Select** is combined with a change to the Multiaddr format,
+  advertising the secure channel protocol through the latter instead of
+  negotiating them in-band. Thus [Downgrade attack]s are no longer possible at
+  the protocol negotiation level and a man-in-the-middle can no longer detect
+  a connection being used for libp2p traffic through the negotation process.
+
+- **Connection establishment**
+
+  In addition to making us vulnerable to downgrade attacks, negotiating the
+  security protocol takes one round-trip in the common case with **[Multistream
+  Select]**. On top of that negotiating a stream multiplexer (on TCP) takes
+  another round-trip.
+
+  **Protocol Select** on the other hand depends on security protocols being
+  advertised, thereby eliminating the need for negotiating them. For optimized
+  implementations, stream muxer negotiation will take zero round-trips for the
+  client (depending on the details of the cryptographic handshake protocol). In
+  that case, the client will be able to immediately open a stream after
+  completing the cryptographic handshake. In addition the protocol supports
+  zero-round-trip optimistic stream protocol negotiation when proposing a single
+  protocol.
+
+- **Data schema**
+
+  The **[Multistream Select]** protocol is defined as a plaintext protocol
+  with no strict schema definition, making both implementation and protocol
+  evolution time consuming and error-prone. See [rust-libp2p/1795] showcasing
+  complexity for implementors and [specs/196] to showcase difficulty evolving
+  protocol.
+
+  The **Protocol Select** protocol will use a binary data format defined in a
+  machine parseable schema language allowing protocol evolution at the schema
+  level.
+
+- **Bandwidth**
+
+  **Multistream Select** is not as bandwidth efficient as it could be. For
+  example negotiating a protocol requires sending the protocol name back and
+  forth. For human readability protocol names are usually long strings (e.g.
+  `/ipfs/kad/1.0.0`).
+
+  **Protocol Select** will include the option to improve bandwidth efficiency
+  e.g. around protocol names in the future. While _Protocol Select_ will not
+  solve this in the first iteration, the protocol is designed with this
+  optimization in mind, and allows for a smooth upgrade in a future iteration.
```

Comment:
Hm. So, I gave this as one of the primary motivations for Protocol Select, but it's not actually specific to Protocol Select, in it's current incarnation.

That is, I could use the same trick of aliasing codec-based protocol names in multistream.

Let's think _very carefully_ about our upgrade path here. I'm now less convinced it's something we can punt because, well, it's kind of the main reason we wanted this protocol in the first place.